### ISSUE TITLE
amp-carousel causing document height to fluctuate / be too large

### ISSUE THREAD

**Issue by brandondiamond** [ampproject/amphtml#11929]
## What's the issue?

We are attempting to measure the document height and scroll position to compute a scroll depth percentage. At the moment, we're using the maximum height of the document over the course of a session.

Unfortunately, it seems that for many documents containing an amp-carousel component, the height can be very large. It can also be somewhat varied.

To mitigate, we could avoid using the maximum height and instead use the last height. But as the height appears to "stabilize" over time, short sessions might still be associated with an invalid, large height.

We chatted with Justin about some possible resolutions and he had a few interesting proposals. I'll let Justin jump in with some extra context.

## How do we reproduce the issue?

Please contact bdiamond@ for more details about this issue.

## What browsers are affected?

All browsers

## Which AMP version is affected?

All versions

**Comment by camelburrito** [comment]
@jridgewell can you give me  more context

**Comment by jridgewell** [comment]
This isn't actually an issue with `<amp-carousel>`, but with `layout="responsive"` and multiple large content inside (ie, the same bug as https://github.com/ampproject/amphtml/issues/9397).

The issue is actually with the layout CSS (so, framework stuff). It's readily apparent when preloading pages such as the Forbes article on https://www.google.com/search?q=site%3Aforbes.com+most-annoying-business-jargon. If you insert breakpoints on the `documentHeight` messages, we can get:

```
$('amp-carousel')
> <amp-carousel class=​"i-amphtml-layout-responsive i-amphtml-layout-size-defined" height=​"900" i-amphtml-layout=​"responsive" layout=​"responsive" type=​"slides" width=​"360">​…​</amp-carousel>​
$('amp-carousel').getBoundingClientRect()
> DOMRect {x: 16, y: 110, width: 288, height: 37408.25, top: 110, …}

// After waiting, carousel gets connected to document.
$('amp-carousel').getBoundingClientRect()
> DOMRect {x: 16, y: 110, width: 288, height: 720, top: 110, …}
```

The issue is because we wait till `#connectedCallback` to apply the `.i-amphtml-notbuilt` class. Once that gets added, the element snaps down from 37k pixels to 720. This is because of:

```css
.i-amphtml-notbuilt:not(.i-amphtml-layout-container)>* {
    display: none;
}
```

This cases all the images inside the carousel to collapse down until the carousel is built. Once it is built, all the images come back to life inside a carousel with appropriate dimensions.

Note that this affects prerendering much more frequently than origin docs, because the browser prioritizes connecting custom element in origin mode. But it _can_ affect origin mode.

- - -

My solution is to do a query selector for `[layout="responsive"]` (and maybe `fill`?), and set the not-built class on the elements if they're still not connected (they have a `Resource` if they've connected). This will fix any issues with `documentHeight`, and it'll doubly fix #9397.

**Comment by dvoytenko** [comment]
@jridgewell Do you mean we add the following to amp.css?

```
[layout="responsive"]:not(.i-amphtml-built)>*:not([placeholder]) {
  display: none !important;
}
```

or such?

(This assumes we introduce `i-amphtml-built` class)

**Comment by jridgewell** [comment]
That could work, too.

**Comment by brandondiamond** [comment]
Hey all! Just finishing up our data analysis doc and wanted to follow up. Is there a fix being prepared for this? Thanks!

**Comment by jridgewell** [comment]
I have a fix in #12077. Using the page from https://github.com/ampproject/amphtml/issues/11929#issuecomment-342283528, I'm able to confirm the carousel images no longer cause 37k heights to be sent.
